### PR TITLE
Moving Wavefront to Sleuth

### PR DESCRIPTION
**Pull request by marcingrzejszczak**

### CODE REVIEW

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontAutoConfigurationTests.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+			service = (StringUtils.hasText(application)) ? service : defaultServiceName(environment);
+			ApplicationTags.Builder builder = new ApplicationTags.Builder(application, service);
+			builder.cluster(environment.getProperty("wavefront.application.cluster"));
+			builder.shard(environment.getProperty("wavefront.application.shard"));
+			return builder.build();
+		}
+
+		private String defaultServiceName(Environment environment) {
+			String applicationName = environment.getProperty("spring.application.name");
+			return (StringUtils.hasText(applicationName)) ? applicationName : "unnamed_service";
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontAutoConfigurationTests.java:202]: Would not this be enough (there is another one in `createFromProperties` and in the next test class too)?
```suggestion
			return environment.getProperty("spring.application.name", "unnamed_service");
```

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+		assertThat(spanRecord.parents).extracting(UUID::toString)
+				.containsExactly("00000000-0000-0000-0000-000000000003");
+		assertThat(spanRecord.followsFrom).isNull();
+		// This tests that RPC spans do not share the same span ID
+		assertThat(spanRecord.spanId.toString()).isNotEqualTo("00000000-0000-0000-0000-000000000003")
+				.matches("^[0-9a-f]{8}\\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\\b[0-9a-f]{12}$");
+		assertThat(spanRecord.name).isEqualTo("GET /api/fn/{id}");
+
+		// spot check the unit is valid (millis not micros)
+		long currentTime = System.currentTimeMillis();
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:94]: Isn't the `startMillis` come from `nanoTime()` originally? Is so, would not it be better to get the `currentTime` from there too?
```suggestion
		long currentTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
```

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+	}
+
+	/** Helps ensure test bugs don't result in hung tests! */
+	<R> R takeRecord(BlockingDeque<R> queue) {
+		R result;
+		try {
+			result = queue.poll(3, TimeUnit.SECONDS);
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:175]: Wasn't it already interrupted (and the interrupt status is cleared)? Why does this needed?

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+		@RequestMapping("/api/fn/{id}")
+		public ResponseEntity<String> fn(@PathVariable("id") String id) {
+			return new ResponseEntity<>(id, HttpStatus.OK);
+		}
+
+		@RequestMapping("/error/{api}")
+		public ResponseEntity<Void> error(@PathVariable("api") String api) {
+			switch (api) {
+			case "brave":
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:322]: Is this indented correctly (switch and case are on the same level)?

**Review by jonatan-ivanov — CHANGES_REQUESTED** [review]

File: spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthBraveSpanHandler.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import brave.handler.MutableSpan;
+import brave.handler.SpanHandler;
+import brave.propagation.TraceContext;
+
+import org.springframework.cloud.sleuth.brave.bridge.BraveFinishedSpan;
+import org.springframework.cloud.sleuth.brave.bridge.BraveTraceContext;
+
+class WavefrontSleuthBraveSpanHandler extends SpanHandler implements Runnable, Closeable {
+
+	final WavefrontSleuthSpanHandler spanHandler;
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthBraveSpanHandler.java:31]: That's a copy from the previous solution. Let me polish it...

File: spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+				50000, // TODO: maxQueueSize should be a property, ya?
+				wavefrontSender, meterRegistry, wavefrontConfig.source(), applicationTags,
+				wavefrontProperties.getRedMetricsCustomTagKeys());
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass({ Tracer.class, TracingCustomizer.class, SpanHandler.class })
+	static class BraveCustomizerConfiguration {
+
+		@Bean(BEAN_NAME)
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java:72]: That's a copy from the previous solution. Let me polish it...

File: spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+	static final String BEAN_NAME = "wavefrontTracingCustomizer";
+
+	@Bean
+	@ConditionalOnBean({ MeterRegistry.class, WavefrontConfig.class, WavefrontSender.class, ApplicationTags.class })
+	WavefrontSleuthSpanHandler wavefrontSleuthSpanHandler(MeterRegistry meterRegistry, WavefrontSender wavefrontSender,
+			ApplicationTags applicationTags, WavefrontConfig wavefrontConfig, WavefrontProperties wavefrontProperties) {
+		return new WavefrontSleuthSpanHandler(
+				// https://github.com/wavefrontHQ/wavefront-opentracing-sdk-java/blob/f1f08d8daf7b692b9b61dcd5bc24ca6befa8e710/src/main/java/com/wavefront/opentracing/reporting/WavefrontSpanReporter.java#L54
+				50000, // TODO: maxQueueSize should be a property, ya?
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java:63]: That's a copy from the previous solution. Let me polish it...

File: spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass({ MeterRegistry.class, WavefrontConfig.class, WavefrontSender.class })
+@ConditionalOnBean({ WavefrontSender.class, ApplicationTags.class })
+@AutoConfigureAfter(WavefrontMetricsExportAutoConfiguration.class)
+@AutoConfigureBefore(BraveAutoConfiguration.class)
+@EnableConfigurationProperties(WavefrontProperties.class)
+@ConditionalOnProperty(value = { "spring.sleuth.enabled", "spring.sleuth.wavefront.enabled" }, matchIfMissing = true)
+public class WavefrontSleuthAutoConfiguration {
+
+	static final String BEAN_NAME = "wavefrontTracingCustomizer";
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java:55]: That's a copy from the previous solution. Let me polish it...

File: spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontSleuthAutoConfiguration.java:72]: That's a copy from the previous solution. Let me polish it...

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java (hunk 3 above)
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:175]: That's a copy from the previous solution.

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java (hunk 4 above)
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:322]: That's a copy from the previous solution. Let me polish it...

Comment (marcingrzejszczak) [reply spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:322]: So it's the autoformatter that does that apparently

File: spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java [hunk 9]
```diff
@@ -91,7 +91,7 @@ void sendsToWavefront() {
 		assertThat(spanRecord.name).isEqualTo("GET /api/fn/{id}");
 
 		// spot check the unit is valid (millis not micros)
-		long currentTime = System.currentTimeMillis();
+		long currentTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java:94]: @jonatan-ivanov I'm reverting this change cause the tests are failing

**Comment by marcingrzejszczak** [comment]
After talking to Stephane Nicoll I'm closing this PR in favour of https://github.com/spring-cloud/spring-cloud-sleuth/pull/1852. The Wavefront-Sleuth integration will remain in Wavefront for the time being. You can read more about this here https://github.com/wavefrontHQ/wavefront-spring-boot/pull/104

**Pull request closed without merging**